source/attribute/AttributeAccess.h

```diff
@@ -0,0 +1,100 @@
+*/
+
+#ifndef ATTRIBUTE_ACCESS_H_
+#define ATTRIBUTE_ACCESS_H_
+
+#include "AttributeCategory.h"
+#include "AttributeEffectType.h"
+
+#include <optional>
+
+class AttributeAccess {
+public:
+	AttributeAccess(const AttributeCategory category, const AttributeEffectType effect);
+	AttributeAccess(const AttributeCategory category, const AttributeEffectType categoryEffect,
+			const AttributeEffectType effect);
+
+	// Accessors
+	AttributeCategory Category() const;
+	AttributeEffectType Effect() const;
+
+	// Checks whether this attribute is a multiplier.
+	bool IsMultiplier() const;
+	static bool IsMultiplier(const AttributeEffectType effect);
+	// Creates a multiplier for this attribute, if not already a multiplier.
+	AttributeAccess Multiplier() const;
+	static AttributeEffectType Multiplier(const AttributeEffectType effect);
+	// Checks whether this attribute is relative.
+	bool IsRelative() const;
+	static bool IsRelative(const AttributeEffectType effect);
+	// Creates a relative version of this attribute, if not already relative.
+	AttributeAccess Relative() const;
+	static AttributeEffectType Relative(const AttributeEffectType effect);
+
+	// Gets the attribute's category effect (variant), or -1 if none.
+	AttributeEffectType GetCategoryEffect() const;
+	static AttributeEffectType GetCategoryEffect(const AttributeCategory category);
+	// Creates a version of this attribute that has the specified effect in its category.
+	AttributeAccess WithCategoryEffect(const AttributeEffectType type) const;
+	static AttributeCategory WithCategoryEffect(const AttributeCategory category, const AttributeEffectType effect);
+
+	// Checks whether this effect is a requirement for its category.
+	// Required effects mark resource consumption when an action is taken.
+	bool IsRequirement() const;
+	static bool IsRequirement(const AttributeCategory category, const AttributeEffectType effect);
+	// Checks if this effect, when used with the PASSIVE category, denotes a capacity or a
+	// passively applied effect.
+	static bool IsCapacity(const AttributeEffectType type);
+	// Gets the basic effect of an attribute category. This is the effect used when
+	// the category is used in a node with a value directly applied to it.
+	static std::optional<AttributeEffectType> GetBaseEffect(const AttributeCategory category);
+
+	// Gets the default minimum value for this attribute.
+	double GetDefaultMinimum() const;
+
+	template <typename A>
+	bool operator==(const A &other) const;
+
+	bool operator<(const AttributeAccess other) const;
+
+private:
+	// Checks if the given attribute category is always composite. These categories are always merged with their effect
+	// in the constructor.
+	static bool IsAlwaysComposite(const AttributeCategory category);
+
+private:
+	// The category and effect type of the described attribute.
+	AttributeCategory category;
+	AttributeEffectType effect;
+};
+
+
+
+template<class T>
+inline bool AttributeAccess::operator==(const T &other) const
```

Comment:
Does this mean that attributes can now be compared to anything in code, and then just get compiled to nonsense?

Reply:
Resolved; handled by my other comment.